Data model and service interfaces

joaoandremartins opened the pull request:
This is the first stab at the data model and service interfaces of Google Foreign Exchange.

There are still some chunks left (Wallet/Venmo integrations, logging and auditing, etc.), but this should do as a base.

Each microservice could then live in com.google.gfx.services, with their own poms, etc.

joaoandremartins committed:
Data model and service interfaces

joaoandremartins committed:
Adds constructors and removes useless id field,

meltsufin commented on src/main/java/com/google/gfx/model/Location.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Used to localise users.
+ */
+public enum Location {
```
Perhaps make it a class rather than an enum. We may want to pull the list of locations from some service instead of hard-coding them.

joaoandremartins replied:
Indeed, this was also on my mind.

meltsufin commented on src/main/java/com/google/gfx/model/Match.java:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * The final transactioned amount as part of this trade.
+   */
+  private float amount;
```
We [shouldn't use float to represent money](http://stackoverflow.com/questions/3730019/why-not-use-double-or-float-to-represent-currency) due to rounding errors. The same goes for the fx rate.

meltsufin commented on src/main/java/com/google/gfx/model/Match.java:
```diff
@@ -0,0 +1,51 @@
+   */
+  private float amount;
+
+  public Match(Order buyOrder, Order sellOrder, float fxRateBuyToSell, Boolean buyerResponse,
```
The constructor should only take required params. I think buyer/seller response are optional.

joaoandremartins replied:
The builder pattern might actually suit the Match class the best. Only one of buyOrder and sellOrder need to be provided at Match creation time.

meltsufin commented on src/main/java/com/google/gfx/model/Match.java:
```diff
@@ -0,0 +1,51 @@
+ * A match to an existing order, generated by a user responding to a buy/sell order or by an cron
+ * task that matches existing orders periodically.
+ */
+public class Match {
```
add `id` as well?

joaoandremartins replied:
Done.

meltsufin commented on src/main/java/com/google/gfx/model/Order.java:
```diff
@@ -0,0 +1,78 @@
+  /**
+   * Currency being sought.
+   */
+  private Currency targetCurrency;
```
I think the name and comment are a little ambiguous. I think `paymentCurrency` or `baseCurrency` is a more clear name.

joaoandremartins replied:
paymentCurrency is also ambiguous, I think.

I still think fromCurrency and toCurrency are pretty self-explanatory. What do you think?

meltsufin commented on src/main/java/com/google/gfx/model/User.java:
```diff
@@ -0,0 +1,28 @@
+  /**
+   * Where the user is located.
+   */
+  private Location location;
```
I think we also talked about adding a "home currency", but that can actually live in the `Location` class.

joaoandremartins replied:
What were you thinking of using home currency for?

meltsufin commented on src/main/java/com/google/gfx/services/exchangerate/ExchangeRateService.java:
```diff
@@ -0,0 +1,23 @@
+   * @param toCurrency gotten currency
+   * @return the exchange rate value, such that 'itself * fromCurrency = toCurrency'
+   */
+  float getExchangeRate(Currency fromCurrency, Currency toCurrency);
```
return `BigDecimal`.

joaoandremartins replied:
Changed amounts and fxRates from float to BigDecimal.

meltsufin commented on src/main/java/com/google/gfx/services/order/OrderService.java:
```diff
@@ -0,0 +1,85 @@
+  /**
+   * Stores an order.
+   */
+  void storeOrder(Order order);
```
`createOrder`? or is this create/update?

joaoandremartins replied:
storeOrder might be preferable since the order is created when you construct the Order object?

aslo commented on src/main/java/com/google/gfx/services/order/OrderService.java:
```diff
@@ -0,0 +1,85 @@
+   * {@code currency}
+   */
+  // TODO(joaomartins): Implement pagination.
+  Collection<Order> listBuyOrdersByAmount(float amount, Currency currency, int numOrders);
```
I think you'd want to return a `List<Order>` here since this method implies ordered results. Same for `listSellOrdersLowerByAmount`

joaoandremartins replied:
Done.

aslo commented on src/main/java/com/google/gfx/services/matching/MatchingService.java:
```diff
@@ -0,0 +1,38 @@
+   *
+   * @param match proposed match
+   */
+  void acceptMatch(Match match);
```
I think you'd probably want some way for the caller to say whether the buyer or seller is accepting this Match, since both acceptances are tracked separately. 

More generally, it might be good to think about adding some sort of a Context object to all of these service methods, to provide some general context data about the caller, for all method calls. This will also be necessary to do authorization. (i.e. user A shouldn't be able to accept an Match between users B and C). 

```
class Context {
  User user;
  // etc, more fields to come later, i.e. auth scopes
}

void acceptMatch(Context ctx, Match match)
```

meltsufin replied:
An alternative to the context could simply be a request object. `MatchDecision` that has user, decision, and match.

joaoandremartins replied:
One thing that jumped to sight was that some of these methods should probably belong to the objects. For instance, acceptMatch() should probably be Match.accept(), instead. Or they could both exist simultaneously, e.g., MatchingService controls whether a match can be decided on by a user and then Match.accept() just registers the decision (or even Match.setBuyerResponse()/Match.setSellerResponse()).

I can't make my mind up on what's best: whether to check if a user can decide on a match inside the acceptMatch() method or outside..

meltsufin replied:
In a service-oriented architecture, it's common to just use dumb data transfer objects that not have methods other than getters and setters, and the services do all the manipulation of the state and business logic.

meltsufin commented on src/main/java/com/google/gfx/services/matching/MatchingService.java:
```diff
@@ -0,0 +1,38 @@
+   * @param order original posted order
+   * @param amount amount of the match, which might fully or partially match the original order
+   */
+  void matchOrder(Order order, float amount);
```
Shouldn't this return a collection of matches? Also, I don't think passing the amount is necessary.

joaoandremartins replied:
What should this method do, in your opinion?

I was thinking it creates a Match for an order, for a certain amount, since you can have a fractional match.

Why would a collection of matches be returned?

meltsufin replied:
You can have several fractional matches that fulfill one order. The other consideration is whether the service is stateless. You might want to publish a "match" event instead of just storing it in the db.

joaoandremartins replied:
Not exactly sure what you mean here. Maybe we can chat when you've got a sec :)

joaoandremartins committed:
Turns Location into a class.
Makes amounts and exchange rates BigDecimals due to float rounding issues.
OrderService.list methods return lists instead of collections, for ordering.

joaoandremartins committed:
Adds emailAddresses to Location.

joaoandremartins commented:
I'm also taking some caution in not going too far deep into technicalities here (e.g., builder pattern), since there's a change the model will change considerably, depending on what datastore choices are available.

e.g., if we can only use a non-relational database (through the Spring JDBC connector), then the object schema will not map directly into database tables, since the object schema wouldn't be fully normalized (e.g., some objects contain List members).

joaoandremartins committed:
Created LocationMailingList.
Added builders to Match and Order.

joaoandremartins commented on src/main/java/com/google/gfx/services/matching/MatchingService.java:
```diff
@@ -0,0 +1,70 @@
+   * This method was conceived to be ran as part of a cron job that matches existing buy and sell
+   * orders.
+   */
+  void matchExistingBuyAndSellOrders();
```
I'm still a bit divided about whether to keep this, but will do for now so this doesn't fall out of the radar.

joaoandremartins committed:
Reorganises project structure so that model is in its own module and all services are in the same module.

joaoandremartins committed:
Sets packaging=jar in new service modules.

joaoandremartins committed:
New lines in pom.xml.

joaoandremartins commented:
@aslo @meltsufin sorry guys, but I had to dramatically change the organization here. But I think all your comments were addressed, so this is probably ready for another go when you have a chance.

The current architecture looks like:
gfx
  - model
  - services
    - discovery
    - exchangerate
    - mail
    - matching
    - order
    - web

meltsufin commented:
Is discovery, essentially the service registry?
Would it make sense to make mail, notification instead, so that we don't lock it into just doing email? Maybe hangouts notification or something?

joaoandremartins commented:
Yes, service registry should be handled by the discovery module.
Is it weird if the service is called notification but it only contains a `sendMail` method? We can't really do away with this method either, so it has to live somewhere.

joaoandremartins committed:
Renames to GfxWebApplication

joaoandremartins committed:
Renames parent pom to services and adds name.

joaoandremartins committed:
Some more tidbit fixes.

joaoandremartins committed:
Removes unnecessary pom stuff.

meltsufin commented on model/src/main/java/com/google/gfx/Match.java:
```diff
@@ -0,0 +1,246 @@
+    this.sellOrder = sellOrder;
+  }
+
+  public void setBuyerResponse(boolean buyerResponse) {
```
`boolean` -> `Boolean` so response can be cleared?

joaoandremartins replied:
Hmm go from responded to unresponded? Sure, sounds like we might need it

joaoandremartins committed:
boolean -> Boolean
